File: api/src/main/java/xtdb/query/QueryOpts.java
package xtdb.query;

import clojure.lang.*;
import xtdb.api.TransactionKey;

import java.time.Duration;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class QueryOpts implements ILookup, Seqable {
    private static final Keyword ARGS_KEY = Keyword.intern("args");
    private static final Keyword BASIS_KEY = Keyword.intern("basis");
    private static final Keyword AFTER_TX_KEY = Keyword.intern("after-tx");
    private static final Keyword TX_TIMEOUT_KEY = Keyword.intern("tx-timeout");
    private static final Keyword DEFAULT_TZ_KEY = Keyword.intern("default-tz");
    private static final Keyword EXPLAIN_KEY = Keyword.intern("explain?");
    private static final Keyword KEY_FN_KEY = Keyword.intern("key-fn");

    private Map<String, Object> args;
    private Basis basis;
    private TransactionKey afterTx;
    private Duration txTimeout;
    private ZoneId defaultTz;
    private boolean explain;
    private String keyFn;

    public QueryOpts(Map<String, Object> args, Basis basis, TransactionKey afterTx, Duration txTimeout,
                        ZoneId defaultTz, boolean explain, String keyFn) {
        this.args = args;
        this.basis = basis;
        this.afterTx = afterTx;
        this.txTimeout = txTimeout;
        this.defaultTz = defaultTz;
        this.explain = explain;
        this.keyFn = keyFn;
    }

    public Map<String, Object> args() {
        return args;
    }

    public Basis basis() {
        return basis;
    }

    public TransactionKey afterTx() {
        return afterTx;
    }

    public Duration txTimeout() {
        return txTimeout;
    }

    public ZoneId defaultTz() {
        return defaultTz;
    }

    public boolean explain() {
        return explain;
    }

    public String keyFn() {
        return keyFn;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        QueryOpts queryOpts = (QueryOpts) o;
        return Objects.equals(args, queryOpts.args) &&
                Objects.equals(basis, queryOpts.basis) &&
                Objects.equals(afterTx, queryOpts.afterTx) &&
                Objects.equals(txTimeout, queryOpts.txTimeout) &&
                Objects.equals(defaultTz, queryOpts.defaultTz) &&
                Objects.equals(explain, queryOpts.explain) &&
                Objects.equals(keyFn, queryOpts.keyFn);
    }

    @Override
    public int hashCode() {
        return Objects.hash(args, basis, afterTx, txTimeout, defaultTz, explain, keyFn);
    }

    @Override
    public String toString() {
        return "QueryRequest{" +
                "args=" + args +
                ", basis=" + basis +
                ", afterTx=" + afterTx +
                ", txTimeout=" + txTimeout +
                ", defaultTz=" + defaultTz +
                ", explain=" + explain +
                ", keyFn=" + keyFn +
                '}';
    }

    @Override
    public Object valAt(Object key) {
        return valAt(key, null);
    }

    @Override
    public Object valAt(Object key, Object notFound) {
        if (key == ARGS_KEY) {
            return args;
        } else if (key == BASIS_KEY) {
            return basis;
        } else if (key == AFTER_TX_KEY){
            return afterTx;
        } else if (key == TX_TIMEOUT_KEY) {
            return txTimeout;
        } else if (key == DEFAULT_TZ_KEY) {
            return defaultTz;
        } else if (key == EXPLAIN_KEY) {
            return explain;
        } else if (key == KEY_FN_KEY) {
           return keyFn;
        } else {
            return notFound;
        }
    }

    @Override
    public ISeq seq() {
        List<Object> seqList = new ArrayList<>();
        if (args != null) {
            seqList.add(MapEntry.create(ARGS_KEY, args));
        }
        if (basis != null) {
            seqList.add(MapEntry.create(BASIS_KEY, basis));
        }
        if (afterTx != null) {
            seqList.add(MapEntry.create(AFTER_TX_KEY, afterTx));
        }
        if (txTimeout != null) {
            seqList.add(MapEntry.create(TX_TIMEOUT_KEY, txTimeout));
        }
        if (defaultTz != null) {
            seqList.add(MapEntry.create(DEFAULT_TZ_KEY, defaultTz));
        }
        if (EXPLAIN_KEY != null) {
            seqList.add(MapEntry.create(EXPLAIN_KEY, explain));
        }
        if (keyFn != null) {
            seqList.add(MapEntry.create(KEY_FN_KEY, keyFn));
        }
        return PersistentList.create(seqList).seq();
    }
}

File: api/src/main/kotlin/xtdb/query/QueryOpts.kt
package xtdb.query

import clojure.lang.*
import xtdb.api.TransactionKey
import java.time.Duration
import java.time.ZoneId
import java.util.ArrayList

private val ARGS_KEY: Keyword = Keyword.intern("args")
private val BASIS_KEY: Keyword = Keyword.intern("basis")
private val AFTER_TX_KEY: Keyword = Keyword.intern("after-tx")
private val TX_TIMEOUT_KEY: Keyword = Keyword.intern("tx-timeout")
private val DEFAULT_TZ_KEY: Keyword = Keyword.intern("default-tz")
private val EXPLAIN_KEY: Keyword? = Keyword.intern("explain?")
private val KEY_FN_KEY: Keyword = Keyword.intern("key-fn")

data class QueryOpts(
    @JvmField val args: Map<String, *>? = null,
    @JvmField val basis: Basis? = null,
    @JvmField val afterTx: TransactionKey? = null,
    @JvmField val txTimeout: Duration? = null,
    @JvmField val defaultTz: ZoneId? = null,
    @JvmField val explain: Boolean = false,
    @JvmField val keyFn: String? = null
) : ILookup, Seqable {

    override fun valAt(key: Any?): Any? {
        return valAt(key, null)
    }

    override fun valAt(key: Any?, notFound: Any?): Any? {
        return when {
            key === ARGS_KEY -> args
            key === BASIS_KEY -> basis
            key === AFTER_TX_KEY -> afterTx
            key === TX_TIMEOUT_KEY -> txTimeout
            key === DEFAULT_TZ_KEY -> defaultTz
            key === EXPLAIN_KEY -> explain
            key === KEY_FN_KEY -> keyFn
            else -> notFound
        }
    }

    override fun seq(): ISeq? {
        val seqList: MutableList<Any?> = ArrayList()
        seqList.add(MapEntry.create(ARGS_KEY, args))

        basis?.let { seqList.add(MapEntry.create(BASIS_KEY, it)) }

        if (afterTx != null) {
            seqList.add(MapEntry.create(AFTER_TX_KEY, afterTx))
        }
        if (txTimeout != null) {
            seqList.add(MapEntry.create(TX_TIMEOUT_KEY, txTimeout))
        }
        if (defaultTz != null) {
            seqList.add(MapEntry.create(DEFAULT_TZ_KEY, defaultTz))
        }

        seqList.add(MapEntry.create(EXPLAIN_KEY, explain))
        seqList.add(MapEntry.create(KEY_FN_KEY, keyFn))

        return PersistentList.create(seqList).seq()
    }

    companion object {
        @JvmStatic
        fun queryOpts() = Builder()
    }

    class Builder {
        private var args: Map<String, *>? = null
        private var basis: Basis? = null
        private var afterTx: TransactionKey? = null
        private var txTimeout: Duration? = null
        private var defaultTz: ZoneId? = null
        private var explain: Boolean = false
        private var keyFn: String? = null

        fun args(args: Map<String, *>?) = apply { this.args = args }
        fun basis(basis: Basis?) = apply { this.basis = basis }
        fun afterTx(afterTx: TransactionKey?) = apply { this.afterTx = afterTx }
        fun txTimeout(txTimeout: Duration?) = apply { this.txTimeout = txTimeout }
        fun defaultTz(defaultTz: ZoneId?) = apply { this.defaultTz = defaultTz }
        fun explain(explain: Boolean) = apply { this.explain = explain }
        fun keyFn(keyFn: String?) = apply { this.keyFn = keyFn }

        fun build() = QueryOpts(args, basis, afterTx, txTimeout, defaultTz, explain, keyFn)
    }
}
